### PR TITLE
allow dataset alias to add more than one dataset events

### PR DESCRIPTION
**Pull request by Lee-W**
## Why
https://github.com/apache/airflow/issues/34206#issuecomment-2344108361

```python
@task(outlets=[DatasetAlias("example-alias")])
def produce_dataset_events(*, outlet_events):
    outlet_events["example-alias"].add(Dataset("dbt-cloud-table-1"))
    outlet_events["example-alias"].add(Dataset("dbt-cloud-table-2"))
```

could be confusing if only the second one create a dataset event

## What
Save all the datasets when `add` is invoked. 

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/utils/context.py [hunk 1]
```diff
@@ -172,7 +172,7 @@ class OutletEventAccessor:
 
     raw_key: str | Dataset | DatasetAlias
     extra: dict[str, Any] = attrs.Factory(dict)
-    dataset_alias_event: DatasetAliasEvent | None = None
+    dataset_alias_events: list[DatasetAliasEvent] = attrs.Factory(factory=list)
```
Comment (uranusjr) [inline airflow/utils/context.py]:
```suggestion
    dataset_alias_events: list[DatasetAliasEvent] = attrs.field(factory=list)
```

Autocomplete fail?

File: airflow/utils/context.py [hunk 2]
```diff
@@ -190,12 +190,11 @@ def add(self, dataset: Dataset | str, extra: dict[str, Any] | None = None) -> No
...
-        if extra:
-            self.extra = extra
-
-        self.dataset_alias_event = DatasetAliasEvent(
-            source_alias_name=dataset_alias_name, dest_dataset_uri=dataset_uri
+        event = DatasetAliasEvent(
+            source_alias_name=dataset_alias_name, dest_dataset_uri=dataset_uri, extra=extra or {}
         )
+        if event not in self.dataset_alias_events:
+            self.dataset_alias_events.append(event)
```
Comment (uranusjr) [inline airflow/utils/context.py]: We should be more efficient than this. A dict or set would be better.

File: airflow/serialization/serialized_objects.py [hunk 3]
```diff
@@ -285,15 +285,17 @@ def encode_outlet_event_accessor(var: OutletEventAccessor) -> dict[str, Any]:
...
 
 
 def decode_outlet_event_accessor(var: dict[str, Any]) -> OutletEventAccessor:
-    raw_key = BaseSerialization.deserialize(var["raw_key"])
-    outlet_event_accessor = OutletEventAccessor(extra=var["extra"], raw_key=raw_key)
-    outlet_event_accessor.dataset_alias_event = var["dataset_alias_event"]
+    outlet_event_accessor = OutletEventAccessor(
+        extra=var["extra"],
+        raw_key=BaseSerialization.deserialize(var["raw_key"]),
+        dataset_alias_events=var["dataset_alias_events"],
```
Comment (uranusjr) [inline airflow/serialization/serialized_objects.py]: We might need some compatibility code for the old `dataset_alias_event` field?

**Review by uranusjr — COMMENTED** [review]
General idea looks good to me, some notes on implementation.

File: airflow/utils/context.py (hunk 1 above)
Comment (Lee-W) [reply airflow/utils/context.py]: Oops... just fixed it. Tanks!

File: airflow/serialization/serialized_objects.py (hunk 3 above)
Comment (Lee-W) [reply airflow/serialization/serialized_objects.py]: sounds good! just added it.

File: airflow/utils/context.py (hunk 2 above)
Comment (Lee-W) [reply airflow/utils/context.py]: Yep, I thought of it. But that might mean we need to make `dataset_alias_events` (`list[dict[str | dict[str, Any]]]`)hashable, which I guess (?) we don't really want to. Another thing I'm thinking of is just adding it anyway and processing them in [taskinstance](https://github.com/apache/airflow/blob/3bc2d3aef6c348e26775f4711ac593acfac9595e/airflow/models/taskinstance.py#L2896) all at once. WDYT?

Comment (Lee-W) [reply airflow/utils/context.py]: We just had an offsite sync-up. We'll let taskinstance do the duplication

**Review by uranusjr — APPROVED** [review]

**Comment by KhushalSh** [comment]
Can someone show me an example usage for this ?
@Lee-W

**Pull request merged**